### PR TITLE
4038: Add Database specific changes

### PR DESCRIPTION
**Pull request by elder-joseph**
## Description
As a developer, I need to refactor the version archiver to handle data that is nested in a parent / child relationship so that additional enhancements can be made.

https://github.com/department-of-veterans-affairs/va.gov-team/issues/4038

These are the Database Specific Changes only.

App and Database changes for 4038 can be found here:
https://github.com/department-of-veterans-affairs/gibct-data-service/pull/550

## Testing done
All CI tests pass locally, along with dev testing of functionality.

## Screenshots
N/A

## Acceptance criteria
- [x] Archiver is configurable to archive tables with version id.
- [x] Archiver is configurable to archive tables modeled as children to tables with version id.
- [x] Archiver moves previous versioned data to configured corresponding archive tables.

## Definition of done
- [x] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: db/migrate/20200107150001_create_school_certifying_officials_archives.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class CreateSchoolCertifyingOfficialsArchives < ActiveRecord::Migration[5.2]
+    def up
+      safety_assured do
+        execute "create table school_certifying_officials_archives (like school_certifying_officials
+          including defaults
+          including constraints
+          including indexes
```
Comment (zurbergram) [inline db/migrate/20200107150001_create_school_certifying_officials_archives.rb]: don't include indexes

**Review by zurbergram — DISMISSED** [review]
You are going to want to update the timestamp in filename to be after most recent migration that is in master

https://github.com/department-of-veterans-affairs/gibct-data-service/blob/master/db/migrate/20200124135649_add_completion_date_to_uploads.rb

**Comment by jholton** [comment]
> You are going to want to update the timestamp in filename to be after most recent migration that is in master

Not strictly necessary unless there's is some conflict with the migrations.  `db:migrate` will see that these haven't been run, and will run them even though they don't have the most recent timestamp(s) in the file name.

File: db/migrate/20200107150000_add_institution_columns_to_school_certifying_officials.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class AddInstitutionColumnsToSchoolCertifyingOfficials < ActiveRecord::Migration[5.2]
+  def change
+    add_column :school_certifying_officials, :institution_id, :integer
```
Comment (jholton) [inline db/migrate/20200107150000_add_institution_columns_to_school_certifying_officials.rb]: Are you setting up a relationship?
If so you might want to use:
```ruby
disable_ddl_transaction!

def change
  add_reference :school_certifying_officials, :institution, foreign_key: true, index: false
  add_index :school_certifying_officials, :institution_id, algorithm: :concurrently
end
```

File: db/migrate/20200107150001_create_school_certifying_officials_archives.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+      );"
+      end
+    end
+  
+     def down
+      drop_table :school_certifying_officials_archives
+    end
+  end
```
Comment (jholton) [inline db/migrate/20200107150001_create_school_certifying_officials_archives.rb]: Nitpick:  got some whitespace issues.

File: db/migrate/20200107150004_remove_institutions_version_constraint.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+
+class RemoveInstitutionsVersionConstraint < ActiveRecord::Migration[5.2]
+    disable_ddl_transaction!
+
+    def change
+        change_column_null :institutions, :version, true
+    end
+  end
```
Comment (jholton) [inline db/migrate/20200107150004_remove_institutions_version_constraint.rb]: Nitpick:  whitespace issues.

**Review by jholton — APPROVED** [review]

**Pull request merged**